### DISCUSSION TITLE
Displaying an uploaded image in the view page

### DISCUSSION THREAD

**Discussion by judgej** [filamentphp/filament#6670] [Help] [answered]
I think I am missing some concept here, but I'll try to explain what I have, and see if there are any pointers towards how to solve this.

I am using the https://github.com/Nuhel/filament-cropper package, which gives me a `Cropper` component. That component extends the `FileUpload` built-in component, by offering some cropping features on the front end.

The features work fine, except that the file that has already been uploaded is not displayed. I would like to display it.

So, with the component in a form, I am using the following method to save the uploaded file:

    saveUploadedFileUsing(function (BaseFileUpload $component, TemporaryUploadedFile $file) {
        ...
    }

I do this because there are a number of steps. When editing a `User`, a signature image can be uploaded:

1. The temporary uploaded file is moved to a local file and renamed with a name that is compatible with S3 conventions.
2. The file is then given to the Spatie Media package that handles copying the file to S3, creating a local media record, and attaching the media record to the `User` model.
3. The local uploaded files are then tidied up.

That bit works - the uploaded signature, after cropping and rotating etc. is now in S3, and it all links to the `User` through a Spatie Media Library record.

If I want the URL of the signature, it is in `$user->media->first()->getTemporaryUrl()`.

So, I don't know if the `filament-cropper` package is supposed to show the uploaded image or images. I'm guessing it is, since it extends the `FileUpload` which does display the images (as seen in the Filament demo app). Maybe it doesn't, and I need to extend the component's view to show it. But I suspect this is not the problem.

In the view for the `filament-cropper` package, if I `@dump($getState())` I see the state of the `User` fields, but the state for the `signature` field is just an empty array. I don't think the uploaded file URL is getting into the view. And why should it? I surely need to tell it how to get the URL.

So that - I think - is my question. How do I tell the `filament-cropper` package, when displaying in a form, that the URL for the previously-uploaded image can be found at  `$user->media->first()->getTemporaryUrl()`? I just don't see where all this stuff connects together.

**Comment by judgej** [comment]
The component is defined as:

```php
Nuhel\FilamentCropper\Components\Cropper::make('signature')
    -> ...options, validation, rules etc...
    ->saveUploadedFileUsing(function (BaseFileUpload $component, TemporaryUploadedFile $file) {
        Save the file to a new name, pass it to Spatie (goes into S3) and attaches to the `User` model
        given by $component->getModelInstance()
    }
```

The preview of the uploaded image works fine, after it can been cropped in the popup:

![image](https://github.com/filamentphp/filament/assets/395934/4fe8f1f9-943c-4839-ac48-f2ff072afb48)

But coming back to the page or refreshing after saving, it is just a blank upload field again:

![image](https://github.com/filamentphp/filament/assets/395934/daab0257-f13c-4b0b-9d4b-62459f55d345)

**Comment by judgej** [comment]
The `User` model does not have a `signature` column. Would I maybe need to create a fake `signature` getter to provide the state that the component can use? If so, what would that state look like? An array with a public URL in it? An object of some kind? A string?

**Comment by danharrin** [comment]
I appreciate this is a problem for you, but this really isn't something the Filament core team can solve. Our built-in FileUpload component works fine in both of these cases, this must be a problem with that third party plugin.

FWIW, in v3 our file upload component will have a cropper built-in.

**Comment by judgej** [comment]
I tried switching to the built-in, and it behaves much the same. It's not that I want someone else to solve it for me, it's more than I'm looking for the understanding on how these components ae supposed to work, where to dig through to find out what is going wrong.

The use case seems to be "files will be uploaded to a filesystem". I found `saveUploadedFileUsing()` which changes that to "files will be given to you to store however you like", and that seems to work. What I can't see is how to tell the component (Cropper or built-in) what it needs to know to display the image. Does it maybe *only* support showing a preview from a Laravel filesystem? Or are there hooks where that can be diverted?

**Reply by judgej** [reply]
Oh, and builtin Cropper - awesome.

**Reply by danharrin** [reply]
Would it be possible to put together a screen recording explaining what is going wrong? I'm not sure I follow what's going wrong here

**Reply by danharrin** [reply, marked as answer]
We do have a Spatie Media library plugin, have you tried that instead of overriding stuff like `saveUploadedFileUsing()`?

**Comment by judgej** [comment]
Sorry, been caught up in other things all day. Will try the Spatie plugin and will go for that as an interim if it works, to look forward to V3 with the cropping. I did get the original `saveUploadedFileUsing()` from the Spatie plugin, so if that works, I can start discecting it to see how the flows work.

**Reply by danharrin** [reply]
Are you using an admin panel resource or the form builder on its own

**Reply by judgej** [reply]
It's in an admin panel, if that's what you mean.

I tried the Spatie field, and apart from the lack of manual cropping, it seemed to work. It was important to set `visibility` to `private` in order to generate the temporary urls to preview the image. The Spatie field also didn't honour the default disk set in `media-library.disk_name`, and always fell back to `local` if not explicitly set. Haven't looked into why yet.

**Reply by danharrin** [reply]
We pass the disk name to media library from our own Filament configuration files. This is to ensure consistent behaviour between all fields

**Reply by judgej** [reply]
I see, the Spatie Media Upload field falls back to the Filament default disk (`forms.default_filesystem_disk`) and not the Spatie Media Library default disk (`media-library.disk_name`) if not specified in the form.

**Reply by danharrin** [reply]
Yeah exactly, which is consistent with other form fields, right?

**Comment by Khant-Nyar** [comment]
Hello i want to make a field (backdrop)
it trigger with checkbox or toggle 
use can choose uploading a file or provide image url 

how can i do it 

example 
```php
     Forms\Components\Toggle::make('backdrop_upload_type')
          ->label('Use File Upload for backdrop ?')
          ->reactive(),
      Forms\Components\FileUpload::make('backdrop')
          ->label('Upload File')
          ->visible(fn ($get) => $get('backdrop_upload_type')),
      Forms\Components\TextInput::make('backdrop')
          ->label('Backdrop URL')
          ->visible(fn ($get) => ! $get('backdrop_upload_type')),
```
